pkg/scheduler/dependency_graph.go

```diff
@@ -500,26 +586,109 @@ func (sched *Scheduler) BuildDependencyGraph(options interfaces.DependencyGraphO
-				existingSr := depGraph.graph[key]
+				existingSr := rootContext.graph.graph[key]
 				if existingSr == nil {
-					log.Printf("Adding resource %s to the dependency graph flow %s", key, options.FlowName)
-					depGraph.graph[key] = entry.scheduledResource
+					if !silent {
+						log.Printf("Adding resource %s to the dependency graph flow %s", key, flow.Name)
+					}
+					rootContext.graph.graph[key] = entry.scheduledResource
 				} else {
 					sched.updateContext(existingSr.Context, entry.parentContext, entry.dependency)
 					existingSr.Requires = append(existingSr.Requires, entry.scheduledResource.Requires...)
 					existingSr.RequiredBy = append(existingSr.RequiredBy, entry.scheduledResource.RequiredBy...)
+					existingSr.usedInReplicas = append(existingSr.usedInReplicas, entry.scheduledResource.usedInReplicas...)
 					for metaKey, metaValue := range entry.scheduledResource.Meta {
 						existingSr.Meta[metaKey] = metaValue
 					}
 				}
 			}
 		}
 	}
+	return nil
+}
 
-	for _, value := range depGraph.graph {
-		value.RequiredBy = unique(value.RequiredBy)
-		value.Requires = unique(value.Requires)
+func (sched *Scheduler) composeFinalizer(construction, destruction *DependencyGraph, replicas []client.Replica) func() {
+	replicaMap := map[string]client.Replica{}
+	for _, replica := range replicas {
+		replicaMap[replica.ReplicaName()] = replica
 	}
 
-	return depGraph, nil
+	var destructors []func() bool
+	var failed chan *ScheduledResource
+
+	var resourceCount int
+	for _, depGraph := range [2]*DependencyGraph{construction, destruction} {
+		for _, resource := range depGraph.graph {
+			release := true
+			for _, replicaName := range resource.usedInReplicas {
+				if _, found := replicaMap[replicaName]; !found {
+					release = false
+					break
+				}
+			}
+			if release {
+				resourceCopy := resource
+				f := func() bool {
+					if !resourceCopy.Existing || resourceCopy.Context.graph.graphOptions.AllowDeleteExternalResources {
+						log.Printf("%s flow: Deleting resource %s", resourceCopy.Context.flow.Name, resourceCopy.Key())
+						err := resourceCopy.Delete()
+						if err != nil {
+							log.Println(err)
+							statusError, ok := err.(*errors.StatusError)
+							if !ok || statusError.Status().Reason != unversioned.StatusReasonNotFound {
+								failed <- resourceCopy
+							}
+						}
+						return err == nil
+					} else {
+						log.Printf("%s: Won't delete external resource %s", resourceCopy.Context.flow.Name, resourceCopy.Key())
+						return true
+					}
+				}
+				destructors = append(destructors, f)
+				resourceCount++
+			}
+		}
+	}
+
+	return func() {
+		failed = make(chan *ScheduledResource, resourceCount)
+		defer close(failed)
+		runConcurrently(destructors, sched.concurrency)
+		failedReplicas := map[string]bool{}
+
+	readFailed:
+		for {
+			select {
+			case res := <-failed:
+				for _, replicaName := range res.usedInReplicas {
+					failedReplicas[replicaName] = true
+				}
+			default:
```

Comment:
Won't this always be the case since `failed` is created in line 655 and not passed anywhere it could get anything between 655 and here?